update comparison docs (#614)
TobikoData/sqlmesh: docs/comparisons.md

## dbt
[dbt](https://www.getdbt.com/) is a tool for data transformations. It is a pioneer in this space and has shown how valuable transformation frameworks can be. Although dbt is a fanstastic tool, it has trouble scaling with data and organizational size.

> dbt initially built their product focused on smaller companies with small data. These companies are able to fully refresh their entire warehouse every run, which usually ensures correctness. Fully refreshing your entire warehouse is not cost effective or feasible at many companies. We've designed SQLMesh to work at any data or organizational scale. This allows data teams of all sizes to compute pipelines correctly without excessives costs or wait times.

Reviewer: This is conjecture of their intentions. Can we say something more from our perspective of what the tool was like at launch?

When dbt arrived on the scene, they brought best practices from software development to the data community, that acceled in environments where the warehouse could be fully refreshed with each run.

> dbt initially built their product focused on smaller companies with small data. These companies are able to fully refresh their entire warehouse every run, which usually ensures correctness. Fully refreshing your entire warehouse is not cost effective or feasible at many companies. Incremental models in DBT is considered an "advanced" feature and leaves a lot of room for error. SQLMesh actively encourages incremental modeling and aims to make it as easy as full refresh.

Reviewer 2: ```suggestion
dbt initially built their product focused on smaller companies with small data. These companies are able to fully refresh their entire warehouse every run, which usually ensures correctness. Fully refreshing your entire warehouse is not cost effective or feasible at many companies. Incremental models in dbt are considered an "advanced" feature and leave a lot of room for error. SQLMesh actively encourages incremental modeling and aims to make it as easy as full refresh.
```

| `Notebook Support`                | ❌ | ✅
| `Comprehensive Python API`        | ❌ | ✅

### Environments

> Development and staging environments in DBT are expensive to make and not fully representative of what will go into production.

Reviewer 2: ```suggestion
Development and staging environments in dbt are expensive to make and not fully representative of what will go into production.
```

> The usual flow for creating a new environment in dbt is to re-run your entire warehouse in a new environment. This may work at small scales, but even if it does, it's a waste of time and money. SQLMesh is able to provide efficient isolated environments with [Virtual Data Marts](concepts/plans.md#plan-application). Creating a development environment in SQLMesh is free -- you can quickly get a full replica of any other environment with a simple command. Environments in dbt cost compute and storage.

Reviewer 2: ```suggestion
The usual flow for creating a new environment in dbt is to rerun your entire warehouse in a new environment. This may work at small scales, but even if it does, it's a waste of time and money. SQLMesh is able to provide efficient isolated environments with [Virtual Data Marts](concepts/plans.md#plan-application). Creating a development environment in SQLMesh is free -- you can quickly get a full replica of any other environment with a simple command. Environments in dbt cost compute and storage.
```

> Additionally, SQLMesh ensures that promotion of staging environments to produciton is predictable and consistent. Promotions are simple pointer swaps meaning there is again no wasted compute. There is no concept of promotion in dbt, and queries are all rerun when it's time to deploy something.

Reviewer 2: ```suggestion
Additionally, SQLMesh ensures that promotion of staging environments to production is predictable and consistent. Promotions are simple pointer swaps meaning there is again no wasted compute. There is no concept of promotion in dbt, and queries are all rerun when it's time to deploy something.
```

The subqueries that look for MAX(date) could have a performance impact on the query. SQLMesh is able to avoid these extra subqueries.

Additionally, dbt expects an incremental model to be able to fully refresh the first time it runs. For some large scale data sets, this is cost prohibitive or infeasible. SQLMesh is able to [batch](../concepts/models/overview#batch_size) up backfills into more manageable chunks.

### SQL understanding
dbt heavily relies on [Jinja](https://jinja.palletsprojects.com/en/3.1.x/). It has no understanding of SQL and treats all queries as raw strings with no context. This means that simple syntax errors (like a trailing comma) are difficult to debug and require a full run to detect.

Although SQLMesh supports Jinja, it does not rely on it and parses/understands SQL through [SQLGlot](https://github.com/tobymao/sqlglot). Simple errors can be detected at compile time. You no longer have to wait minutes to see that you've referenced a column incorrectly or missed a comma.

> Additionally, having a first-class understanding of SQL allows for SQLMesh to do some interesting things, like transpilation, column-level lineage, and automatic change categorization.

Reviewer 2: ```suggestion
Additionally, having a first-class understanding of SQL allows SQLMesh to do some interesting things, like transpilation, column-level lineage, and automatic change categorization.
```

### Testing

> dbt calls data quality checks testing. Although data quality checks are extremely valuable, they are not sufficient for creating robust data pipelines. Data qualiy checks are great for detecting upstream data issues and large scale problems like nulls and duplicates. But they are not meant for testing edge cases or business logic.

Reviewer 2: ```suggestion
dbt calls data quality checks testing. Although data quality checks are extremely valuable, they are not sufficient for creating robust data pipelines. Data quality checks are great for detecting upstream data issues and large scale problems like nulls and duplicates. But they are not meant for testing edge cases or business logic.
```
